lib/internal/test_runner/mock/mock_timers.js

```diff
@@ -0,0 +1,343 @@
+const nodeTimers = require('timers');
+const nodeTimersPromises = require('timers/promises');
+const EventEmitter = require('events');
+
+function compareTimersLists(a, b) {
+  return (a.runAt - b.runAt) || (a.id - b.id);
+}
+
+function setPosition(node, pos) {
+  node.priorityQueuePosition = pos;
+}
+
+function abortIt(signal) {
+  return new AbortError(undefined, { cause: signal.reason });
+}
+
+const SUPPORTED_TIMERS = ['setTimeout', 'setInterval'];
+
+class MockTimers {
+  #realSetTimeout;
+  #realClearTimeout;
+  #realSetInterval;
+  #realClearInterval;
+
+  #realPromisifiedSetTimeout;
+  #realPromisifiedSetInterval;
+
+  #realTimersSetTimeout;
+  #realTimersClearTimeout;
+  #realTimersSetInterval;
+  #realTimersClearInterval;
+
+  #timersInContext = [];
+  #isEnabled = false;
+  #currentTimer = 1;
+  #now = DateNow();
+
+  #executionQueue = new PriorityQueue(compareTimersLists, setPosition);
+
+  #setTimeout = FunctionPrototypeBind(this.#createTimer, this, false);
+  #clearTimeout = FunctionPrototypeBind(this.#clearTimer, this);
+  #setInterval = FunctionPrototypeBind(this.#createTimer, this, true);
+  #clearInterval = FunctionPrototypeBind(this.#clearTimer, this);
+
+  constructor() {
+    emitExperimentalWarning('The MockTimers API');
+  }
+
+  #createTimer(isInterval, callback, delay, ...args) {
+    const timerId = this.#currentTimer++;
+    this.#executionQueue.insert({
+      __proto__: null,
+      id: timerId,
+      callback,
+      runAt: DateNow() + delay,
+      interval: isInterval,
+      args,
+    });
+
+    return timerId;
+  }
+
+  #clearTimer(position) {
+    this.#executionQueue.removeAt(position);
+  }
+
+  async * #setIntervalPromisified(interval, startTime, options) {
+    const context = this;
+    const emitter = new EventEmitter();
+    if (options?.signal) {
+      validateAbortSignal(options.signal, 'options.signal');
+
+      if (options.signal?.aborted) {
+        throw abortIt(options.signal);
+      }
+
+      const onabort = (reason) => {
+        emitter.emit('data', { aborted: true, reason });
+        options?.signal?.removeEventListener('abort', onabort);
+      };
+
+      options.signal?.addEventListener('abort', onabort);
+    }
+
+    const eventIt = EventEmitter.on(emitter, 'data');
+    const callback = () => {
+      startTime += interval;
+      emitter.emit('data', startTime);
+    };
+
+    const timerId = this.#createTimer(true, callback, interval, options);
+
+    const iterator = {
+      __proto__: null,
+      [SymbolAsyncIterator]() {
+        return this;
+      },
+      async next() {
+        const result = await eventIt.next();
+        const value = ArrayPrototypeAt(result.value, 0);
+        if (value?.aborted) {
+          iterator.return();
+          throw abortIt(options.signal);
+        }
+
+        return {
+          __proto__: null,
+          done: result.done,
+          value,
+        };
+      },
+      async return() {
+        emitter.removeAllListeners();
```

Comment:
Remove the abort listener here too?

Also, we need to implement `throw()` too if this isn't a generator for compatibiility

Reply:
got it thanks! For the abort listener, I'll follow what @MoLow  suggested here https://github.com/nodejs/node/pull/47775#discussion_r1218556093

just implemented throw